src/simulation.py

```diff
@@ -0,0 +1,233 @@
+        probabilities = array([1 / (2 ** (1 + x)) for x in range(num)])
+
+        # Normalize the probabilities to 1
+        n_probabilities = probabilities / np.sum(probabilities)
+        self.probabilities = n_probabilities
+        return n_probabilities
+
+    def norm_opti_probs(
+        self,
+    ):
+        """
+        Create a uniform distribution with mg elements, where mg is the max
+        generations / max number of polygons
+        """
+        self.probabilities = [1 / self.max_generations] * self.max_generations
+        return self.probabilities
+
+    def eval_loss(self, image):
+        """
+        Evaluate an image to the base_image and return the SAD
+        """
+        return sad(self.base_image, image)
+
+    def cc_loss(self, parent, child):
+        """
+        Compute and compare loss
+        """
+
+        # compute the loss of the child iteration with the parent
+        l_parent = self.eval_loss(parent)
+        l_child = self.eval_loss(child)
+
+        # compare loss
+        if l_child < l_parent:
+            self.counter = 0
+        else:
+            self.counter += 1
+
+        return l_parent, l_child
+
+    def select(
+        self,
+    ):
+        """
+        Using probabilities, randomly select and return a polygon from the
+        canvas sequence and its index
+        """
+        indx, self.polygon_i = np.random.choice(
+            list(enumerate(self.canvas.sequence)), p=self.probabilities
+        )
+
+        return indx, self.polygon_i
+
+    def create_polygon(self, c: Canvas):
+        """
+        Create polygon and add it to the canvas
+        """
+
+        c = add_polygon(canvas=c, polygon=polygon_init(id=self.canvas.how_many()))
+
+        self.counter = 0
+        self.update_probabilities()
+        return c
+
+    def run(
+        self,
+    ):
+        """
+        Run the simulation until completion.
+        """
+
+        # initialize vars
+        t = 0
+
+        generations = []
+
+        newer_solution = None
+        older_solution = None
+        # get loss of current solution
+        v_k = self.eval_loss(self.canvas.image())
+
+        while t <= self.num_evals:
+            # TODO: make this a method
+            _indx, self.polygon_i = self.select()
+
+            # use temporary variables to store previous and current solutions
+            older_solution = copy(self.canvas)
+            newer_solution, child = polygon_mutate(self.canvas, self.polygon_i)
+
+            # compare and compute the child with the parent loss
+            l_parent, l_child = self.cc_loss(older_solution, newer_solution)
+
+            t += 1
+
+            if (self.counter > self.max_iterations) and (
+                self.canvas.how_many() < self.max_generations
+            ):
+                if l_child < v_k:
+                    # update the canvas to the improved version
+                    generations.append(deepcopy(self.canvas))
```

Comment:
We need to accept the `child` or `newer_solution` into the `self.canvas` before appending here since this will be unchanged. We probably should modify `mutate_polygon` so that it returns a copy of a canvas that already has the mutated polygon incorporated.

Reply:
This has been implemented -- mutate polygon now returns a copy of the canvas with the mutations applied